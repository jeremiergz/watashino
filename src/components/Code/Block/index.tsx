import Text, { TextProps } from 'components/primitives/Text';
import React, { forwardRef } from 'react';

const BlockCode = forwardRef<HTMLElement, BlockCodeProps>(({ children, ...rest }, ref) => {
  return (
    <Text as="code" fontSize={16} ref={ref} textAlign="left" {...rest}>
      {children}
    </Text>
  );
});

BlockCode.displayName = 'BlockCode';

export type BlockCodeProps = TextProps;
export default BlockCode;
