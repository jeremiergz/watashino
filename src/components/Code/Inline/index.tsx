import Text, { TextProps } from 'components/primitives/Text';
import { useTheming } from 'components/providers/ThemeProvider';
import React from 'react';

const InlineCode: React.FC<InlineCodeProps> = ({ children, ...rest }) => {
  const { theme } = useTheming();
  return (
    <Text
      as="code"
      backgroundColor={theme.type === 'dark' ? 'lightgray' : 'vsCodeDark'}
      borderRadius={8}
      color={theme.type === 'dark' ? 'black' : 'white'}
      display="inline-flex"
      fontSize={16}
      paddingX={2}
      paddingY="2px"
      textAlign="left"
      {...rest}
    >
      {children}
    </Text>
  );
};

InlineCode.displayName = 'InlineCode';

export type InlineCodeProps = TextProps;
export default InlineCode;
