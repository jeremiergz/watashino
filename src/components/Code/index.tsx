import Button from 'components/Button';
import BlockCode from 'components/Code/Block';
import InlineCode from 'components/Code/Inline';
import FlexBox from 'components/primitives/FlexBox';
import Text, { TextProps } from 'components/primitives/Text';
import CopyIcon from 'components/svgs/icons/Copy';
import copyToClipboard from 'copy-to-clipboard';
import React, { useRef, useState } from 'react';

let timeout: number;

const Code: React.FC<CodeProps> = ({ children, className, ...rest }) => {
  const [hasCopied, setHasCopied] = useState(false);
  const codeRef = useRef<HTMLPreElement>();
  const handleCopyClick = () => {
    if (codeRef && codeRef.current) {
      copyToClipboard(codeRef.current.textContent);
      setHasCopied(true);
      clearTimeout(timeout);
      timeout = setTimeout(() => setHasCopied(false), 2500);
    }
  };
  const isBlockCode = !!className;
  return isBlockCode ? (
    <>
      <BlockCode ref={codeRef} {...rest}>
        {children}
      </BlockCode>
      <FlexBox as="span" alignItems="center" left={3} marginTop={4} position="sticky" width={0}>
        <Button onClick={handleCopyClick} padding={2} variant="text">
          <CopyIcon fill="white" height={16} width={16} />
          <Text color="white" marginLeft={1}>
            {hasCopied ? 'copied!' : 'copy'}
          </Text>
        </Button>
      </FlexBox>
    </>
  ) : (
    <InlineCode {...rest}>{children}</InlineCode>
  );
};

Code.displayName = 'Code';

export type CodeProps = TextProps;
export default Code;
