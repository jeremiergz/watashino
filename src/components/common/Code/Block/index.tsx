import React, { forwardRef } from 'react';
import Text, { TextProps } from '../../../primitives/Text';

type BlockCodeProps = TextProps;

const BlockCode = forwardRef<HTMLElement, BlockCodeProps>(({ children, ...rest }, ref) => {
  return (
    <Text as="code" fontSize={16} ref={ref} textAlign="left" {...rest}>
      {children}
    </Text>
  );
});

BlockCode.displayName = 'BlockCode';

export { BlockCodeProps };
export default BlockCode;
