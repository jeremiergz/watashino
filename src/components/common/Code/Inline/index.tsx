import React from 'react';
import Text, { TextProps } from '../../../primitives/Text';

type InlineCodeProps = TextProps;

const InlineCode = ({ children, ...rest }: InlineCodeProps) => {
  return (
    <Text
      as="code"
      backgroundColor="secondary"
      borderRadius={8}
      color="white"
      display="inline-block"
      fontSize={16}
      paddingBottom={1}
      paddingTop="2px"
      paddingX={2}
      textAlign="left"
      verticalAlign="bottom"
      {...rest}
    >
      {children}
    </Text>
  );
};

InlineCode.displayName = 'InlineCode';

export { InlineCodeProps };
export default InlineCode;
