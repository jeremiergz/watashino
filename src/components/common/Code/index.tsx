import copyToClipboard from 'copy-to-clipboard';
import React, { useRef, useState } from 'react';
import Flex from '../../primitives/Flex';
import Text, { TextProps } from '../../primitives/Text';
import Copy from '../../svg/Copy';
import Button from '../Button';
import BlockCode from './Block';
import InlineCode from './Inline';

type CodeProps = TextProps;

let timeout: number;

const Code = ({ children, className, ...rest }: CodeProps) => {
  const [hasCopied, setHasCopied] = useState(false);
  const codeRef = useRef<HTMLPreElement>();
  const handleCopyClick = () => {
    if (codeRef && codeRef.current) {
      copyToClipboard(codeRef.current.textContent);
      setHasCopied(true);
      clearTimeout(timeout);
      timeout = setTimeout(() => setHasCopied(false), 2500);
    }
  };
  const isBlockCode = !!className;
  return (
    <>
      {isBlockCode ? (
        <BlockCode ref={codeRef} {...rest}>
          {children}
        </BlockCode>
      ) : (
        <InlineCode {...rest}>{children}</InlineCode>
      )}
      <Flex as="span" alignItems="center" left={3} marginTop={2} position="sticky" width={0}>
        <Button onClick={handleCopyClick} padding={2}>
          <Copy fill="white" height={24} width={24} />
          <Text color="white" marginLeft={1}>
            {hasCopied ? 'copied!' : 'copy'}
          </Text>
        </Button>
      </Flex>
    </>
  );
};

Code.displayName = 'Code';

export { CodeProps };
export default Code;
