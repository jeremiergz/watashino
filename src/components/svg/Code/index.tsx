import React from 'react';
import Svg, { SvgProps } from '../../common/Svg';
import { useTheming } from '../../core/ThemingManager';

type CodeProps = SvgProps;

const Code = (props: CodeProps) => {
  const { theme } = useTheming();
  return (
    <Svg viewBox="0 0 512 512" {...props}>
      <path
        fill={theme.colors.primary}
        d="M0 256C0 114.8 114.313.3 255.441 0c141.246-.3 255.414 112.652 256.551 253.89.613 76.227-32.097 144.825-84.43 192.11C382.165 404.984 322 380 256 380c-64.14 0-122.77 23.594-167.687 62.57-2.25 1.953-5.614 1.868-7.782-.175C30.957 395.707 0 329.473 0 256zm0 0"
      ></path>
      <path
        fill="#98e5fb"
        d="M429.07 149.07a10 10 0 012.93 7.07v285.75C386.121 485.345 324.176 512 256 512s-130.121-26.656-176-70.11V70.11C91.836 58.897 86.191 48.804 100 40h215.86a10 10 0 017.07 2.93zm0 0"
      ></path>
      <path
        fill="#c7eefd"
        d="M429.879 150H327a5 5 0 01-5-5V42.121c.324.25.637.516.93.809l106.14 106.14c.293.293.559.606.809.93zm0 0"
      ></path>
      <path
        fill="#62dafa"
        d="M118 40c-5.523 0-10 4.477-10 10v414.906a257.333 257.333 0 01-28-23.015V50c0-5.523 4.477-10 10-10zm0 0"
      ></path>
      <path
        fill="#028fce"
        d="M200 165v10a5 5 0 01-5 5h-59a5 5 0 01-5-5v-10a5 5 0 015-5h59a5 5 0 015 5zm-35 45h-29a5 5 0 00-5 5v10a5 5 0 005 5h29a5 5 0 005-5v-10a5 5 0 00-5-5zm0 40h-29a5 5 0 00-5 5v10a5 5 0 005 5h29a5 5 0 005-5v-10a5 5 0 00-5-5zm0 80h-29a5 5 0 00-5 5v10a5 5 0 005 5h29a5 5 0 005-5v-10a5 5 0 00-5-5zm0 40h-29a5 5 0 00-5 5v10a5 5 0 005 5h29a5 5 0 005-5v-10a5 5 0 00-5-5zm0 0"
      ></path>
      <path
        fill="#01bff9"
        d="M281 225v-10a5 5 0 015-5h19a5 5 0 015 5v10a5 5 0 01-5 5h-19a5 5 0 01-5-5zm-71 80v-10a5 5 0 00-5-5h-69a5 5 0 00-5 5v10a5 5 0 005 5h69a5 5 0 005-5zm135-15H235a5 5 0 00-5 5v10a5 5 0 005 5h110a5 5 0 005-5v-10a5 5 0 00-5-5zm-149.5-60H265a5 5 0 005-5v-10a5 5 0 00-5-5h-69.5a5 5 0 00-5 5v10a5 5 0 005 5zm0 40H225a5 5 0 005-5v-10a5 5 0 00-5-5h-29.5a5 5 0 00-5 5v10a5 5 0 005 5zm69.5 60h-69.5a5 5 0 00-5 5v10a5 5 0 005 5H265a5 5 0 005-5v-10a5 5 0 00-5-5zm110-80H245.25a5 5 0 00-5 5v10a5 5 0 005 5H375a5 5 0 005-5v-10a5 5 0 00-5-5zm0 0"
      ></path>
      <path
        fill="#1f3595"
        d="M470 480H260c-5.523 0-10-4.477-10-10V350c0-5.523 4.477-10 10-10h210c5.523 0 10 4.477 10 10v120c0 5.523-4.477 10-10 10zm0 0"
      ></path>
      <path
        fill={theme.colors.secondary}
        d="M470 480H280c-5.523 0-10-4.477-10-10V350c0-5.523 4.477-10 10-10h190c5.523 0 10 4.477 10 10v120c0 5.523-4.477 10-10 10zm0 0"
      ></path>
      <path
        fill="#fff5f5"
        d="M451.695 417.07l-35.101 35.102a4.998 4.998 0 01-7.07 0l-5.989-5.988a4.998 4.998 0 010-7.07l27.7-27.7a2 2 0 000-2.828l-27.7-27.695a5.003 5.003 0 010-7.075l5.988-5.988a4.998 4.998 0 017.07 0l35.102 35.102c3.907 3.902 3.907 10.234 0 14.14zm-125.23-36.183a4.998 4.998 0 000-7.07l-5.988-5.989a4.998 4.998 0 00-7.07 0l-35.102 35.102c-3.907 3.902-3.907 10.234 0 14.14l35.101 35.102a4.998 4.998 0 007.07 0l5.989-5.988a4.998 4.998 0 000-7.07l-27.7-27.7a2 2 0 010-2.828zM378.902 360h-9.804a5.002 5.002 0 00-4.903 4.02l-18 90c-.617 3.093 1.75 5.98 4.903 5.98h9.804a5.002 5.002 0 004.903-4.02l18-90c.617-3.093-1.75-5.98-4.903-5.98zm0 0"
      ></path>
    </Svg>
  );
};

Code.displayName = 'Code';

export default Code;
