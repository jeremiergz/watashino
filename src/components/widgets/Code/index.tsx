import React from 'react';
import Text, { TextProps } from '../../primitives/Text';

type CodeProps = TextProps;

const Code = ({ children, ...rest }: CodeProps) => {
    return (
        <Text as="code" fontSize={16} textAlign="left" {...rest}>
            {children}
        </Text>
    );
};

Code.displayName = 'Code';

export { CodeProps };
export default Code;
