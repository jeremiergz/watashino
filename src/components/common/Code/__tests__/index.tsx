import React from 'react';
import ReactTestRenderer from 'react-test-renderer';
import { ThemeProvider } from 'styled-components';
import { DarkTheme } from '../../../../theme';
import Code from '../../Code';

describe('components/common/Code Test Suite', () => {
  it('matches snapshot', () => {
    const blockCodeJSX = (
      <ThemeProvider theme={DarkTheme}>
        <Code className="code">Code Component</Code>
      </ThemeProvider>
    );
    const inlineCodeJSX = (
      <ThemeProvider theme={DarkTheme}>
        <Code>Code Component</Code>
      </ThemeProvider>
    );
    expect(ReactTestRenderer.create(blockCodeJSX).toJSON()).toMatchSnapshot();
    expect(ReactTestRenderer.create(inlineCodeJSX).toJSON()).toMatchSnapshot();
  });
});
