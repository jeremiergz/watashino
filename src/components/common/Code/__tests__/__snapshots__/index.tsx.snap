// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/widgets/Code Test Suite matches snapshot 1`] = `
Array [
  <code
    className="sc-AxirZ idqhbh"
    color="white"
    display="inline-block"
    fontSize={16}
  >
    Code Component
  </code>,
  <span
    className="sc-AxjAm ezCmCO"
    display="flex"
    width={0}
  >
    <button
      className="sc-AxhCb OPdng"
      onClick={[Function]}
    >
      <svg
        className="sc-AxiKw gjbmxY"
        height={24}
        viewBox="0 0 24 24"
        width={24}
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M4 6H2v14c0 1.1.9 2 2 2h14v-2H4V6zm16-4H8c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V4c0-1.1-.9-2-2-2zm-1 9H9V9h10v2zm-4 4H9v-2h6v2zm4-8H9V5h10v2z"
          fill="white"
        />
      </svg>
      <span
        className="sc-AxirZ cdMprA"
        color="white"
      >
        copy
      </span>
    </button>
  </span>,
]
`;
